### PR TITLE
actions/image-partition: allow extended and logical

### PR DESCRIPTION
**Pull request by daichifukui**
This commit allows us to create extended or logical
partitions on MBR partition table (msdos) in order to
address the following issues:

* Because part-type for parted is hardcoded with primary,
we cannot create more than 4 partitions using extended and
logical.

* Even if we manage to create extended and logical partitions,
the verificaton step fails because the Verify function assumes
that the partition number is always sequetially incremental.

This commit makes it possible that if the number of partitions
is more than 4, 3 of them are automatically created as primary,
1 of them is extended, and the remaining partitions are logical.

Signed-off-by: Daniel Sangorrin <daniel.sangorrin@toshiba.co.jp>
Signed-off-by: Daichi Fukui <daichi1.fukui@toshiba.co.jp>

### CODE REVIEW

**Comment by daichifukui** [comment]
This is related to https://github.com/go-debos/debos/issues/293

**Comment by daichifukui** [comment]
Hi.

It's been a while since we began this discussion.
Does anyone have an opinion or feedback on this idea?

File: actions/image_partition_action.go [hunk 1]
```diff
@@ -479,7 +479,17 @@ func (i ImagePartitionAction) Run(context *debos.DebosContext) error {
 		if i.PartitionType == "gpt" {
 			name = p.PartLabel
 		} else {
-			name = "primary"
+			if cap(i.Partitions) <= 4 {
```
Comment (sjoerdsimons) [inline actions/image_partition_action.go:482]:
```suggestion
			if len(i.Partitions) <= 4 {
```
cap is the capacity of the  array; the actual length might be shorter

File: actions/image_partition_action.go [hunk 2]
```diff
@@ -656,6 +666,35 @@ func (i ImagePartitionAction) PostMachineCleanup(context *debos.DebosContext) er
 }
 
 func (i *ImagePartitionAction) Verify(context *debos.DebosContext) error {
+
+	for idx, _ := range i.Partitions {
+		p := &i.Partitions[idx]
+
+		if idx == 3 && cap(i.Partitions) > 4 {
```
Comment (sjoerdsimons) [inline actions/image_partition_action.go:673]:
```suggestion
		if idx == 3 && len(i.Partitions) > 4 {
```
same

File: actions/image_partition_action.go [hunk 3]
```diff
@@ -656,6 +666,35 @@ func (i ImagePartitionAction) PostMachineCleanup(context *debos.DebosContext) er
...
+			var name string
+			var part Partition
+
+			name = "extended"
+			part.number    = idx+1
+			part.Name      = name
+			part.Start     = p.Start
+			tmp_n          := cap(i.Partitions)-1
```
Comment (sjoerdsimons) [inline actions/image_partition_action.go:681]:
```suggestion
			tmp_n          := len(i.Partitions)-1
```

**Review by sjoerdsimons — CHANGES_REQUESTED** [review]

**Review by sjoerdsimons — APPROVED** [review]

**Comment by robertkirkman** [comment]
Hi I don't know enough about the project to know how to create a minimal reproducible example, but when i use [this](https://github.com/foxlet/VaporOS/blob/master/vapor-steamdeck-main.yaml) file with `debos`, if I use commit 03ec3cbfab6a73087ebbb58317f65e117342c1ba of `debos`  it works but if I use commit da773ccbc899d2f5b236f80db303062693f6f2b6 or any of the recent commits after that, the `debos` command gets stuck inside the fakemachine because of a partition named "extended" getting incorrectly created in the wrong place even though the `.yaml` file specifies `gpt` and not `mbr`

**Comment by obbardc** [comment]
@robertkirkman Can you please open a separate issue about this so it doesn't get lost ?

**Comment by robertkirkman** [comment]
ok I opened that here #473

**Pull request merged**